### PR TITLE
get match items

### PR DESCRIPTION
**Pull request by anunayajoshi**
image

### CODE REVIEW

File: backend/matching/src/services/get-match-items.ts [hunk 1]
```diff
@@ -1,7 +1,133 @@
...
+
+  try {
+    if (!userId1 || !userId2) {
+      throw new Error('Both user IDs are required');
+    }
+
+    // Fetch attempted questions for both users
+    const [attemptedQuestions1, attemptedQuestions2] = await Promise.all([
+      fetchAttemptedQuestions(userEndpoint, userId1),
+      fetchAttemptedQuestions(userEndpoint, userId2)
```
Comment (SeeuSim) [inline backend/matching/src/services/get-match-items.ts]: Is it better to fetch a single list given a list of userIds? Saves RTT of 2 requests and lets database handle it on user service end

File: backend/matching/README.md [hunk 2]
```diff
@@ -7,6 +7,9 @@
     docker build \
       -t match-express-local \
       --build-arg port=9004 \
+      --build-arg USER_SERVER_ENDPOINT=http://host.docker.internal:9001 \
+      --build-arg QUESTION_SERVER_ENDPOINT=http://host.docker.internal:9002 \
+      --build-arg COLLAB_SERVICE_ENDPOINT=http://host.docker.internal:9003 \
```
Comment (SeeuSim) [inline backend/matching/README.md]: not necessary - leave this to running with the .env file where necessary

File: backend/matching/.env.local [hunk 3]
```diff
@@ -5,6 +5,11 @@ EXPRESS_PORT=9004
 MATCHING_DB_HOSTNAME="localhost"
 MATCHING_DB_PORT=6379
 
+USER_SERVER_ENDPOINT="http://localhost:9001"
+QUESTION_SERVER_ENDPOINT="http://localhost:9002"
+COLLAB_SERVER_ENDPOINT="http://localhost:9003"
```
Comment (SeeuSim) [inline backend/matching/.env.local]: change names to `PEERPREP_USER_HOST` and import them in `config.ts`

File: backend/matching/src/services/get-match-items.ts [hunk 4]
```diff
@@ -1,7 +1,133 @@
...
+    } else if (searchIdentifier === 'topic' && topic) {
+      payload.topic = topic;
+    } else if (searchIdentifier === 'exact match' && topic && difficulty) {
+      payload.topic = topic;
+      payload.difficulty = difficulty;
+    }
+
+    // Query the question endpoint using the /random endpoint
+    const questionResponse = await axios.post<IServiceResponse<{ question: IQuestion }>>(
+      `${questionEndpoint}/questions/random`,
```
Comment (SeeuSim) [inline backend/matching/src/services/get-match-items.ts]: define all the routes and service clients being called in a separate services file for better maintainability

File: backend/matching/src/services/get-match-items.ts [hunk 5]
```diff
@@ -1,7 +1,133 @@
...
+    );
+
+    if (questionResponse.status !== 200 || !questionResponse.data.data) {
+      throw new Error(questionResponse.data.error?.message || 'Failed to get a random question');
+    }
+
+    const questionId = questionResponse.data.data.question.id;
+
+    // Update attempted questions for both users
+    await Promise.all([
```
Comment (SeeuSim) [inline backend/matching/src/services/get-match-items.ts]: leave this logic to when both users actually connect to the room and complete the interview later on

File: backend/matching/src/workers/matcher.ts [hunk 6]
```diff
@@ -70,7 +74,9 @@ async function processMatch(
       ]);
 
       // Notify both sockets
-      const { ...matchItems } = getMatchItems();
+      const { ...matchItems } = await getMatchItems(searchIdentifier, topic, difficulty, requestorUserId, matchedUserId);
+      logger.info(`matchItems: ${JSON.stringify(matchItems)}`);
```
Comment (SeeuSim) [inline backend/matching/src/workers/matcher.ts]: can place better logging here i.e. Generated Match:

File: backend/user/src/controllers/questions/index.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
+import { StatusCodes } from 'http-status-codes';
+import type { IRouteHandler } from '@/types';
+import { getAttemptedQuestionsService, addAttemptedQuestionService } from '@/services/questions';
+
+export const addAttemptedQuestion: IRouteHandler = async (req, res) => {
+  const userId = req.params.userId; // Assuming the userId is passed as a route parameter
```
Comment (SeeuSim) [inline backend/user/src/controllers/questions/index.ts]: can consider having a list of user ids to assist batch requests by interview room on FE

File: backend/user/src/routes/user.ts [hunk 8]
```diff
@@ -0,0 +1,11 @@
+import express from 'express';
+
+import { getAttemptedQuestions, addAttemptedQuestion } from '@/controllers/questions';
+
+
+const router = express.Router();
+
+router.get('/:userId/attempted-questions/', getAttemptedQuestions)
+router.post('/:userId/attempt-question', addAttemptedQuestion)
```
Comment (SeeuSim) [inline backend/user/src/routes/user.ts]: /attempted-question/get, /attempted-question/add with JSON POST payloads for both as well as support for userId lists

**Review by SeeuSim — CHANGES_REQUESTED** [review]
code in general looks good, but before committing please run the linter to fix code smells:

```sh
# root folder
git add .

npx lint-staged
```

File: backend/matching/src/services/get-match-items.ts (hunk 1 above)
Comment (anunayajoshi) [reply backend/matching/src/services/get-match-items.ts]: ehh, yes. did it this way cos it allows us to use the same endpoint on frontend, to check which questions have been attempted, which we don't have a route for yet, so might as well handle 2 cases at the same time. shdnt be a big deal to leave it as 2 queries.

**Pull request merged**